fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -874,6 +876,104 @@ ACTOR Future<Void> preresolutionProcessing(CommitBatchContext* self) {
 	return Void();
 }
 
+namespace {
+ErrorOr<Optional<std::pair<TenantName, int64_t>>> getTenantDetails(ProxyCommitData* commitData,
```

Comment:
I liked the invariant where `unknown_tenant` is thrown as an error from this method if `tenantId` input is provided; would like to keep it as is, unless we feel strong about it.

Reply:
I don't think this is providing us any benefit here, and we're giving different implementations for two essentially identical paths. The outcome of this function is either 1) we found the tenant, here is its name, or 2) we didn't find the tenant.

And since we are already passing the ID, even that doesn't need to be returned. So in that case, I think this could simplify to a return type of `Optional<TenantName>`. If the optional is empty, then you can infer that the tenant was unknown.

Reply:
If we do want to keep this taking an optional tenant ID rather than a non-optional one (we currently don't use the optional variant, so I don't think we should do this), then we probably should make the case where the tenant ID being invalid also returns an unknown_tenant error rather than an empty optional.

Reply:
I think the preferred signature for this function as it is currently used would be something like:

```
Optional<TenantName> getTenantName(ProxyCommitData *commitData, int64_t tenantId) {}
```

Reply:
In this case, the input `tenantId` is obtained by parsing FDBKey string, so `unknown_tenant` scenario is still possible. Updating the signature not to throw an error means updating the API contract such that either caller gets `required details` OR `empty optional`. 

My suggestion is there could be scenario where caller would rather throw than getting empty optional, which isn't true for the usecase at hand, but, hard to justify would be true for all future scenarios too. 
Thoughts?

Reply:
The purpose of this function seems to be to look up a tenant name based on the passed in ID, and the cleanest way to do that is an optional. You could return an error, but since we are only using one error it serves exactly the same purpose as the optional and is less clear.

If some future caller wants this scenario to turn into an error, then they can certainly do so based on their needs (which we can't really anticipate without knowing the details). They could throw an error if they get an empty optional, or we could even change the signature of the function at that time if that seemed better.

Since we don't have a clear use-case for anything more than the simple version proposed above, I think it's preferable to keep the logic simple.

Reply:
Fair enough, will explicitly document API contract in code documentation to clarify this point. 

Thanks